File: src/sdk/python/ratel_ai/runtime_events.py
"""Public runtime-event stream and catalog snapshot seams (ADR-0019)."""

from __future__ import annotations

import asyncio
import inspect
import json
import os
import secrets
import time
import uuid
from collections.abc import Awaitable, Callable, Coroutine, Sequence
from typing import TYPE_CHECKING, Any, Protocol, cast

from ._native import NativeEventSubscription

if TYPE_CHECKING:
    from .catalog import ToolCatalog
    from .skill_catalog import SkillCatalog

RuntimeEvent = dict[str, Any]
RuntimeEventHandler = Callable[[list[RuntimeEvent]], None | Awaitable[None]]

RUNTIME_EVENT_TYPES = (
    "search",
    "skill_search",
    "gateway_search",
    "invoke_start",
    "invoke_end",
    "invoke_error",
    "gateway_invoke",
    "gateway_error",
    "skill_invoke",
    "index_churn",
    "skill_churn",
    "upstream_register",
    "upstream_invoke",
    "upstream_error",
    "auth_refresh",
    "auth_needs",
    "auth_flow_start",
    "auth_flow_end",
    "experiment_selection",
    "experiment_results",
    "experiment_comparison",
    "experiment_skip",
    "experiment_fallback",
    "experiment_drop",
    "experiment_invocation",
    "experiment_outcome",
    "events_dropped",
)
RUNTIME_EVENT_MAX_PAYLOAD_BYTES = 64 * 1_024
RUNTIME_EVENT_MAX_QUERY_BYTES = 4 * 1_024
RUNTIME_EVENT_MAX_HITS = 100

_REQUIRED_ENVELOPE_FIELDS = {"v", "event_id", "ts", "session_id", "source_id", "type"}


def new_runtime_event_id(now_ms: int | None = None) -> str:
    """Mint a client ULID for one runtime event and OTel projection."""
    alphabet = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
    timestamp = now_ms if now_ms is not None else int(time.time() * 1_000)
    encoded_time = ""
    for _ in range(10):
        encoded_time = alphabet[timestamp % 32] + encoded_time
        timestamp //= 32
    entropy = int.from_bytes(secrets.token_bytes(10), "big")
    encoded_entropy = ""
    for shift in range(75, -1, -5):
        encoded_entropy += alphabet[(entropy >> shift) & 31]
    return encoded_time + encoded_entropy


class _EventSource(Protocol):
    def subscribe_events(
        self,
        handler: Callable[[list[RuntimeEvent]], object],
        *,
        session_id: str,
        source_id: str,
        queue_capacity: int,
        batch_size: int,
    ) -> NativeEventSubscription: ...


class RuntimeEventSubscription:
    """Handle over one merged bounded runtime-event subscription."""

    def __init__(
        self,
        subscriptions: list[NativeEventSubscription],
        pending: set[asyncio.Future[None]],
        active: list[bool],
    ) -> None:
        """Create a subscription over native handles and pending handler work."""
        self._subscriptions = subscriptions
        self._pending = pending
        self._active = active

    async def flush(self) -> None:
        """Wait until accepted native and async-handler work completes."""
        await asyncio.gather(
            *(asyncio.to_thread(subscription.flush) for subscription in self._subscriptions)
        )
        await asyncio.sleep(0)
        while self._pending:
            await asyncio.gather(*tuple(self._pending), return_exceptions=True)
            await asyncio.sleep(0)

    def unsubscribe(self) -> None:
        """Stop accepting new events; already-queued native envelopes still drain."""
        if not self._active[0]:
            return
        self._active[0] = False
        for subscription in self._subscriptions:
            subscription.unsubscribe()

    @property
    def dropped_count(self) -> int:
        """Envelopes displaced by this subscriber's native queues."""
        return sum(subscription.dropped_count for subscription in self._subscriptions)


class RuntimeEvents:
    """Merged push stream over tool and skill catalogs."""

    def __init__(
        self,
        sources: Sequence[_EventSource],
        *,
        session_id: str | None = None,
        source_id: str | None = None,
        queue_capacity: int = 1_024,
        batch_size: int = 64,
    ) -> None:
        """Create a stream sharing one identity across all event sources."""
        self.session_id = session_id or str(uuid.uuid4())
        self.source_id = source_id or _default_source_id()
        self._sources = tuple(sources)
        self._queue_capacity = queue_capacity
        self._batch_size = batch_size

    def subscribe(self, handler: RuntimeEventHandler) -> RuntimeEventSubscription:
        """Subscribe to bounded asynchronous runtime-event batches."""
        active = [True]
        pending: set[asyncio.Future[None]] = set()
        try:
            loop: asyncio.AbstractEventLoop | None = asyncio.get_running_loop()
        except RuntimeError:
            loop = None

        def schedule_awaitable(awaitable: Awaitable[None]) -> None:
            if loop is None:
                close = getattr(awaitable, "close", None)
                if callable(close):
                    close()
                return

            def schedule() -> None:
                future = asyncio.ensure_future(awaitable, loop=loop)
                pending.add(future)
                future.add_done_callback(
                    lambda finished: _finish_async_handler(finished, pending)
                )

            try:
                on_target_loop = asyncio.get_running_loop() is loop
            except RuntimeError:
                on_target_loop = False
            if on_target_loop:
                schedule()
            else:
                loop.call_soon_threadsafe(schedule)

        if inspect.iscoroutinefunction(handler):
            if loop is None:
                raise RuntimeError("async runtime-event handlers require a running event loop")
            async_handler = cast(Callable[[list[RuntimeEvent]], Coroutine[Any, Any, None]], handler)

            def deliver(batch: list[RuntimeEvent]) -> None:
                normalized = [_normalize_runtime_event(event) for event in batch]

                def schedule() -> None:
                    schedule_awaitable(async_handler(normalized))

                loop.call_soon_threadsafe(schedule)

        else:

            def deliver(batch: list[RuntimeEvent]) -> None:
                try:
                    result = handler([_normalize_runtime_event(event) for event in batch])
                    if inspect.isawaitable(result):
                        schedule_awaitable(result)
                except Exception:
                    pass

        subscriptions = [
            source.subscribe_events(
                deliver,
                session_id=self.session_id,
                source_id=self.source_id,
                queue_capacity=self._queue_capacity,
                batch_size=self._batch_size,
            )
            for source in self._sources
        ]
        return RuntimeEventSubscription(subscriptions, pending, active)


def _default_source_id() -> str:
    if service_name := os.getenv("OTEL_SERVICE_NAME"):
        return service_name
    attributes = os.getenv("OTEL_RESOURCE_ATTRIBUTES", "")
    for entry in attributes.split(","):
        key, separator, value = entry.strip().partition("=")
        if separator and key == "service.name" and value:
            return value
    return "ratel"


def _finish_async_handler(
    task: asyncio.Future[None],
    pending: set[asyncio.Future[None]],
) -> None:
    pending.discard(task)
    if task.cancelled():
        return
    task.exception()


def _normalize_runtime_event(event: RuntimeEvent) -> RuntimeEvent:
    normalized = cast(RuntimeEvent, _sanitize_value(event))
    if _serialized_size(normalized) <= RUNTIME_EVENT_MAX_PAYLOAD_BYTES:
        return normalized

    for key in tuple(normalized):
        if key not in _REQUIRED_ENVELOPE_FIELDS and not _is_product_fact_field(key):
            del normalized[key]
            normalized["payload_truncated"] = True
            if _serialized_size(normalized) <= RUNTIME_EVENT_MAX_PAYLOAD_BYTES:
                return normalized

    bounded = {key: normalized[key] for key in normalized if key in _REQUIRED_ENVELOPE_FIELDS}
    bounded["payload_truncated"] = True
    for key, value in normalized.items():
        if key in _REQUIRED_ENVELOPE_FIELDS or not _is_product_fact_field(key):
            continue
        bounded[key] = _sanitize_bounded_value(value)
        if _serialized_size(bounded) > RUNTIME_EVENT_MAX_PAYLOAD_BYTES:
            del bounded[key]
    return bounded


def _sanitize_value(value: Any, key: str = "") -> Any:
    if isinstance(value, str):
        limit = RUNTIME_EVENT_MAX_QUERY_BYTES if key == "query" else 4_096
        return _truncate_utf8(value, limit)
    if isinstance(value, list):
        return [_sanitize_value(item) for item in value[:RUNTIME_EVENT_MAX_HITS]]
    if isinstance(value, dict):
        return {
            str(child_key): _sanitize_value(child_value, str(child_key))
            for child_key, child_value in list(value.items())[:100]
        }
    return value


def _sanitize_bounded_value(value: Any) -> Any:
    if isinstance(value, str):
        return _truncate_utf8(value, 512)
    if isinstance(value, list):
        return [_sanitize_bounded_value(item) for item in value[:RUNTIME_EVENT_MAX_HITS]]
    if isinstance(value, dict):
        return {
            str(key): _sanitize_bounded_value(child)
            for key, child in list(value.items())[:32]
        }
    return value


def _truncate_utf8(value: str, max_bytes: int) -> str:
    encoded = value.encode()
    if len(encoded) <= max_bytes:
        return value
    return encoded[:max_bytes].decode(errors="ignore")


def _serialized_size(value: Any) -> int:
    return len(json.dumps(value, separators=(",", ":")).encode())


def _is_product_fact_field(key: str) -> bool:
    return key.endswith(("_id", "_ids", "_ms", "_count", "_score", "_scores")) or key in {
        "query",
        "target",
        "origin",
        "top_k",
        "hits",
        "outcome",
        "error",
        "error_class",
        "transport",
        "role",
        "cold",
        "agreement",
        "reason",
        "label",
        "score",
        "attributed",
        "rank",
        "turn",
        "action",
    }


class RuntimeCatalog:
    """Complete serializable tool and skill state for snapshot publication."""

    def __init__(
        self,
        tools: ToolCatalog,
        skills: SkillCatalog,
        *,
        source_id: str,
    ) -> None:
        """Create a snapshot view sharing the stream's stable source identity."""
        self._tools = tools
        self._skills = skills
        self._source_id = source_id

    def snapshot(self) -> dict[str, Any]:
        """Return the current full replacement snapshot."""
        return {
            "source_id": self._source_id,
            "tools": self._tools.snapshot(),
            "skills": self._skills.snapshot(),
        }

File: src/sdk/ts/src/runtime-events.ts
import { randomBytes, randomUUID } from "node:crypto";
import type { NativeEventSubscription } from "../native/index.cjs";
import type { ToolDefinition } from "./catalog.js";
import type { SkillDefinition } from "./skill-catalog.js";

const DEFAULT_SOURCE_ID = "ratel";
const REQUIRED_ENVELOPE_FIELDS = new Set([
  "v",
  "event_id",
  "ts",
  "session_id",
  "source_id",
  "type",
]);

/** Frozen remotely publishable v1 event names from ADR-0019. */
export const RUNTIME_EVENT_TYPES = [
  "search",
  "skill_search",
  "gateway_search",
  "invoke_start",
  "invoke_end",
  "invoke_error",
  "gateway_invoke",
  "gateway_error",
  "skill_invoke",
  "index_churn",
  "skill_churn",
  "upstream_register",
  "upstream_invoke",
  "upstream_error",
  "auth_refresh",
  "auth_needs",
  "auth_flow_start",
  "auth_flow_end",
  "experiment_selection",
  "experiment_results",
  "experiment_comparison",
  "experiment_skip",
  "experiment_fallback",
  "experiment_drop",
  "experiment_invocation",
  "experiment_outcome",
  "events_dropped",
] as const;

/** Maximum serialized size of one public event envelope. */
export const RUNTIME_EVENT_MAX_PAYLOAD_BYTES = 64 * 1_024;
/** Maximum UTF-8 byte size of a public search query. */
export const RUNTIME_EVENT_MAX_QUERY_BYTES = 4 * 1_024;
/** Maximum ranked hits carried by one public search event. */
export const RUNTIME_EVENT_MAX_HITS = 100;

/** Stable envelope shared by every public runtime event (ADR-0019). */
export interface RuntimeEvent {
  /** Envelope schema version. */
  readonly v: 2;
  /** Canonical client ULID used to deduplicate and join OTel. */
  readonly event_id: string;
  /** Unix timestamp in milliseconds. */
  readonly ts: number;
  /** Process/session identity shared by the merged stream. */
  readonly session_id: string;
  /** Stable deployment source identity. */
  readonly source_id: string;
  /** Additive event discriminator. */
  readonly type: string;
  /** Lifecycle group for invocation start/end/error events. */
  readonly invocation_id?: string;
  /** Catalog generation when supplied by the producer. */
  readonly catalog_version?: string;
  /** Optional deployment environment. */
  readonly environment?: string;
  /** Optional pseudonymous application user identity. */
  readonly end_user_id?: string;
  /** Active OTel trace identity when a recording span exists. */
  readonly trace_id?: string;
  /** Active OTel span identity when a recording span exists. */
  readonly span_id?: string;
  readonly [field: string]: unknown;
}

/** Identity and bounded-delivery options for {@link RuntimeEvents}. */
export interface RuntimeEventsOptions {
  /** Stable identity for this process/session. Defaults to a fresh UUID. */
  sessionId?: string;
  /** Stable deployment source. Defaults to OTel `service.name`, then `"ratel"`. */
  sourceId?: string;
  /** Per-registry subscriber queue capacity. Defaults to the native bridge default. */
  queueCapacity?: number;
  /** Maximum envelopes per callback batch. Defaults to the native bridge default. */
  batchSize?: number;
}

/** One runtime-events subscription. */
export interface RuntimeEventSubscription {
  /** Stop accepting new events; already-queued native envelopes still drain. */
  unsubscribe(): void;
  /** Wait until accepted registry events reach the handler and async work settles. */
  flush(): Promise<void>;
  /** Envelopes dropped from this subscriber's bounded native queues. */
  readonly droppedCount: number;
}

/** Complete executor-free catalog state published separately from runtime events. */
export interface CatalogSnapshot {
  /** Stable deployment source shared with runtime envelopes. */
  readonly source_id: string;
  /** Complete serializable tool definitions, sorted by id. */
  readonly tools: readonly ToolDefinition[];
  /** Complete public skill metadata, sorted by id and excluding bodies. */
  readonly skills: readonly SkillDefinition[];
}

/** Public catalog-state seam. */
export interface RuntimeCatalog {
  /** Return a current, serializable full replacement snapshot. */
  snapshot(): CatalogSnapshot;
}

/** Asynchronous, fail-open consumer of one public event batch. */
export type RuntimeEventHandler = (batch: readonly RuntimeEvent[]) => void | PromiseLike<void>;

interface SdkSubscriber {
  readonly handler: RuntimeEventHandler;
  readonly pending: Set<Promise<void>>;
}

interface EventSource {
  subscribeEvents(
    handler: (batch: RuntimeEvent[]) => void,
    options: Required<RuntimeEventsOptions>,
  ): NativeEventSubscription;
}

/** Public merged push stream over tool, skill, and SDK-owned runtime facts. */
export class RuntimeEvents {
  /** Process/session identity stamped onto every delivered envelope. */
  readonly sessionId: string;
  /** Stable deployment identity stamped onto events and catalog snapshots. */
  readonly sourceId: string;
  readonly #options: Required<RuntimeEventsOptions>;
  readonly #sources: readonly EventSource[];
  readonly #sdkSubscribers = new Set<SdkSubscriber>();

  /** @internal Constructed by {@link ratel}; consumers use `runtime.events`. */
  constructor(sources: readonly EventSource[], options: RuntimeEventsOptions = {}) {
    this.sessionId = options.sessionId ?? randomUUID();
    this.sourceId = options.sourceId ?? defaultSourceId();
    this.#options = {
      sessionId: this.sessionId,
      sourceId: this.sourceId,
      queueCapacity: options.queueCapacity ?? 1_024,
      batchSize: options.batchSize ?? 64,
    };
    this.#sources = sources;
  }

  /**
   * Subscribe to merged runtime-event batches. Delivery is asynchronous and
   * bounded; handler work never blocks the emitting registry operation.
   */
  subscribe(handler: RuntimeEventHandler): RuntimeEventSubscription {
    const sdkSubscriber: SdkSubscriber = { handler, pending: new Set() };
    const deliver = (batch: RuntimeEvent[]): void => {
      trackHandlerWork(sdkSubscriber, batch.map(normalizeRuntimeEvent));
    };
    const subscriptions = this.#sources.map((source) =>
      source.subscribeEvents(deliver, this.#options),
    );
    this.#sdkSubscribers.add(sdkSubscriber);
    let active = true;
    return {
      unsubscribe: () => {
        if (!active) return;
        active = false;
        this.#sdkSubscribers.delete(sdkSubscriber);
        for (const subscription of subscriptions) subscription.unsubscribe();
      },
      flush: async () => {
        await Promise.all(subscriptions.map((subscription) => subscription.flush()));
        await Promise.all([...sdkSubscriber.pending]);
      },
      get droppedCount() {
        return subscriptions.reduce((total, subscription) => total + subscription.droppedCount, 0);
      },
    };
  }

  /** @internal Merge an SDK-owned fact (experiments) into every public subscriber. */
  emit(event: Record<string, unknown>): RuntimeEvent {
    const eventId = typeof event.event_id === "string" ? event.event_id : newRuntimeEventId();
    const envelope = normalizeRuntimeEvent({
      ...event,
      v: 2,
      event_id: eventId,
      ts: Date.now(),
      session_id: this.sessionId,
      source_id: this.sourceId,
      type: String(event.type ?? "unknown"),
    });
    for (const subscriber of this.#sdkSubscribers) {
      const pending = new Promise<void>((resolve) => {
        queueMicrotask(() => {
          try {
            void Promise.resolve(subscriber.handler([envelope])).then(resolve, resolve);
          } catch {
            resolve();
          }
        });
      });
      subscriber.pending.add(pending);
      void pending.finally(() => subscriber.pending.delete(pending));
    }
    return envelope;
  }
}

function trackHandlerWork(subscriber: SdkSubscriber, batch: readonly RuntimeEvent[]): void {
  let pending: Promise<void>;
  try {
    pending = Promise.resolve(subscriber.handler(batch)).then(
      () => {},
      () => {},
    );
  } catch {
    // Runtime-event consumers are observational and fail open.
    return;
  }
  subscriber.pending.add(pending);
  void pending.then(() => subscriber.pending.delete(pending));
}

function defaultSourceId(): string {
  if (process.env.OTEL_SERVICE_NAME) return process.env.OTEL_SERVICE_NAME;
  const serviceName = process.env.OTEL_RESOURCE_ATTRIBUTES?.split(",")
    .map((entry) => entry.trim().split("=", 2))
    .find(([key]) => key === "service.name")?.[1];
  return serviceName || DEFAULT_SOURCE_ID;
}

/** Minimal monotonicity-free ULID generator: event uniqueness, time sorting, wire alphabet. */
/** @internal Mint the canonical client event id before stream + OTel projection. */
export function newRuntimeEventId(now = Date.now()): string {
  const alphabet = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";
  let time = now;
  let encodedTime = "";
  for (let index = 0; index < 10; index += 1) {
    encodedTime = alphabet[time % 32] + encodedTime;
    time = Math.floor(time / 32);
  }
  const entropy = randomBytes(10);
  let buffer = 0;
  let bits = 0;
  let encodedEntropy = "";
  for (const byte of entropy) {
    buffer = (buffer << 8) | byte;
    bits += 8;
    while (bits >= 5) {
      bits -= 5;
      encodedEntropy += alphabet[(buffer >> bits) & 31];
    }
  }
  return encodedTime + encodedEntropy;
}

function normalizeRuntimeEvent(input: Record<string, unknown>): RuntimeEvent {
  const normalized = sanitizeValue(input, "") as Record<string, unknown>;
  if (serializedSize(normalized) <= RUNTIME_EVENT_MAX_PAYLOAD_BYTES) {
    return normalized as unknown as RuntimeEvent;
  }

  for (const key of Object.keys(normalized)) {
    if (!REQUIRED_ENVELOPE_FIELDS.has(key) && !isProductFactField(key)) {
      delete normalized[key];
      normalized.payload_truncated = true;
      if (serializedSize(normalized) <= RUNTIME_EVENT_MAX_PAYLOAD_BYTES) break;
    }
  }
  if (serializedSize(normalized) <= RUNTIME_EVENT_MAX_PAYLOAD_BYTES) {
    return normalized as unknown as RuntimeEvent;
  }

  const bounded = Object.fromEntries(
    Object.entries(normalized).filter(([key]) => REQUIRED_ENVELOPE_FIELDS.has(key)),
  );
  bounded.payload_truncated = true;
  for (const [key, value] of Object.entries(normalized)) {
    if (REQUIRED_ENVELOPE_FIELDS.has(key) || !isProductFactField(key)) continue;
    bounded[key] = sanitizeBoundedValue(value);
    if (serializedSize(bounded) > RUNTIME_EVENT_MAX_PAYLOAD_BYTES) delete bounded[key];
  }
  return bounded as unknown as RuntimeEvent;
}

function sanitizeValue(value: unknown, key: string): unknown {
  if (typeof value === "string") {
    return truncateUtf8(value, key === "query" ? RUNTIME_EVENT_MAX_QUERY_BYTES : 4_096);
  }
  if (Array.isArray(value)) {
    return value.slice(0, RUNTIME_EVENT_MAX_HITS).map((item) => sanitizeValue(item, ""));
  }
  if (value !== null && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value)
        .slice(0, 100)
        .map(([childKey, childValue]) => [childKey, sanitizeValue(childValue, childKey)]),
    );
  }
  return value;
}

function sanitizeBoundedValue(value: unknown): unknown {
  if (typeof value === "string") return truncateUtf8(value, 512);
  if (Array.isArray(value)) {
    return value.slice(0, RUNTIME_EVENT_MAX_HITS).map((item) => sanitizeBoundedValue(item));
  }
  if (value !== null && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value)
        .slice(0, 32)
        .map(([key, child]) => [key, sanitizeBoundedValue(child)]),
    );
  }
  return value;
}

function truncateUtf8(value: string, maxBytes: number): string {
  if (Buffer.byteLength(value, "utf8") <= maxBytes) return value;
  let result = Buffer.from(value, "utf8").subarray(0, maxBytes).toString("utf8");
  while (Buffer.byteLength(result, "utf8") > maxBytes) result = result.slice(0, -1);
  return result;
}

function serializedSize(value: unknown): number {
  return Buffer.byteLength(JSON.stringify(value), "utf8");
}

function isProductFactField(key: string): boolean {
  return (
    key.endsWith("_id") ||
    key.endsWith("_ids") ||
    key.endsWith("_ms") ||
    key.endsWith("_count") ||
    key.endsWith("_score") ||
    key.endsWith("_scores") ||
    [
      "query",
      "target",
      "origin",
      "top_k",
      "hits",
      "outcome",
      "error",
      "error_class",
      "transport",
      "role",
      "cold",
      "agreement",
      "reason",
      "label",
      "score",
      "attributed",
      "rank",
      "turn",
      "action",
    ].includes(key)
  );
}
